[LoopVectorize] In LoopVectorize.cpp start using getSymbolicMaxBackedgeTakenCount

david-arm opened the pull request:
LoopVectorizationLegality currently only treats a loop as legal to vectorise
if PredicatedScalarEvolution::getBackedgeTakenCount returns a valid SCEV, or
more precisely that the loop must have an exact backedge taken count.
Therefore, in LoopVectorize.cpp we can safely replace all calls to
getBackedgeTakenCount with calls to getSymbolicMaxBackedgeTakenCount, since
the result is the same.

This also helps prepare the loop vectoriser for PR #88385.

david-arm committed:
[LoopVectorize] In LoopVectorize start using getSymbolicMaxBackedgeTakenCount

LoopVectorizationLegality currently only treats a loop as legal
to vectorise if PredicatedScalarEvolution::getBackedgeTakenCount
returns a valid SCEV, or more precisely that the loop must have
an exact backedge taken count. Therefore, in LoopVectorize.cpp
we can safely replace all calls to getBackedgeTakenCount with
calls to getSymbolicMaxBackedgeTakenCount, since the result is
the same.

This also helps prepare the loop vectoriser for PR #88385.

paulwalker-arm reviewed (approved):
I think you might be taking the "create small patches" idea a bit too far but I suppose this change does ensure LoopVectorize is not relying on the more restricted behaviour of `getBackedgeTakenCount()`.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9584,7 +9584,7 @@ static bool processLoopInVPlanNativePath(
     LoopVectorizationRequirements &Requirements) {
 
-  if (isa<SCEVCouldNotCompute>(PSE.getBackedgeTakenCount())) {
+  if (isa<SCEVCouldNotCompute>(PSE.getSymbolicMaxBackedgeTakenCount())) {
```
Does the change the behavior to allow vectorizing loops where getBackedgeTakenCount is SCEVCouldNotCompute? May need a test case and would be an unintended change?

david-arm replied:
You're absolutely right. I was being too eager! I looked and there is no defence against loops with uncountable exits, since LoopVectorizationLegality doesn't filter those out in advance. I've reverted this change.

fhahn replied:
Would it be possible to add a test case?

david-arm replied:
Given my patch no longer changes this code are you asking if I can add a negative test for the native vplan case that proves we don't vectorise outer loops with an early exit? I can have a look, although I'm not sure that's even possible.

david-arm replied:
Anyway, not sure if this is what you wanted or not but I added a test for an outer loop with an early exit. It didn't seem to be covered by any existing tests, although the legality code bails out too early to even attempt calculating a backedge taken count, due to the form of the conditional branch that is rejected by canVectorizeOuterLoop. I hacked the legality code and verified that we can get a symbolic max backedge take count for the outer loop, but not a normal one.

fhahn replied:
>  can add a negative test for the native vplan case that proves we don't vectorise outer loops with an early exit?

yes that was what I had in mind to guard against the regression in the future, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -907,7 +907,7 @@ Value *getRuntimeVF(IRBuilderBase &B, Type *Ty, ElementCount VF) {
 const SCEV *createTripCountSCEV(Type *IdxTy, PredicatedScalarEvolution &PSE,
                                 Loop *OrigLoop) {
-  const SCEV *BackedgeTakenCount = PSE.getBackedgeTakenCount();
+  const SCEV *BackedgeTakenCount = PSE.getSymbolicMaxBackedgeTakenCount();
```
I think this would at least need a comment explaining why this is used here, as at the moment this might be very surprising

fhahn reviewed (commented):
> I think you might be taking the "create small patches" idea a bit too far but I suppose this change does ensure LoopVectorize is not relying on the more restricted behaviour of getBackedgeTakenCount().

Yeah, I think as is this may be more confusing for the reader, at least without a comment. Might also clarify to with an assertion that getBackedgeTakenCount is not SCEVCouldNotCompute....

david-arm commented:
I've added comments explaining why it's ok to use the symbolic max variant. I also fixed rename a variable in createInitialVPlan that was shadowing a variable of the same name in the VPlan class.

paulwalker-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4075,7 +4075,12 @@ LoopVectorizationCostModel::computeMaxVF(ElementCount UserVF, unsigned UserIC) {
         UserIC ? *MaxPowerOf2RuntimeVF * UserIC : *MaxPowerOf2RuntimeVF;
     ScalarEvolution *SE = PSE.getSE();
-    const SCEV *BackedgeTakenCount = PSE.getBackedgeTakenCount();
+    // Currently only loops with countable exits are vectorized so it's safe to
+    // use getSymbolicMaxBackedgeTakenCount as it should give the same result
+    // as getBackedgeTakenCount.
```
To me the comment suggests when uncountable exists are support calling `getSymbolicMaxBackedgeTakenCount` becomes unsafe?, which I think is the opposite of what you mean.

Perhaps something akin to:

Currently only loops with countable exits are vectorized, but calling `getSymbolicMaxBackedgeTakenCount` allows enablement work for loops with uncountable exits whilst also ensuring the symbolic maximum and known back-edge taken count remain identical for loops with countable exits.

david-arm replied:
Seems like a good suggestion! I've rebased the patch to see if the libc++ tests start passing now, and updated the comment. Two for the price of one. :)

david-arm committed:
Address review comment

* Updated comments around calls to getSymbolicMaxBackedgeTakenCount

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4054,7 +4054,13 @@ LoopVectorizationCostModel::computeMaxVF(ElementCount UserVF, unsigned UserIC) {
+    // uncountable exits whilst also ensuring the symbolic maximum and known
+    // back-edge taken count remain identical for loops with countable exits.
+    const SCEV *BackedgeTakenCount = PSE.getSymbolicMaxBackedgeTakenCount();
+    assert(!isa<SCEVCouldNotCompute>(BackedgeTakenCount) &&
```
My original suggestion was along the lines of
```suggestion
    assert(BackedgeTakenCount == PSE.getBackedgeTakenCount() &&
```

david-arm committed:
Address review comments

* Updated assert.
* Add outer loop vectorisation test that has an early exit in the
outer loop. I checked and there are no existing tests for this. It
fails in LoopVectorizationLegality so we never reach the point
in LoopVectorize.cpp where attempt to calculate the backedge taken
count.

david-arm committed:
Re-add "assert(!isa<SCEVCouldNotCompute>(BackedgeTakenCount)"

fhahn reviewed (commented):
LGTM, thanks!

david-arm committed:
Tidy up the new test!
